perf: new Home and Search pages

abruere opened the pull request:
For improved app loading time as mesured by Lightouse after latest update which made our score drop by 30 points or so.
Of course it’s just a set of metrics but it’s also a proxy for performance and loading time perceived by end users.

**Breaking UI changes:**

- Feature columns on home page are hidden unless `VUE_APP_HOME_FEATURES_COLUMNS` is set to true
- `homeHeroBase64` is removed from `styles.json`. SVG background placeholder is now expected to be a plain .svg file named `home-blurred-background.svg` in `src/assets` directory.

**Result:**

We hit 90+ pagespeed score again, with much less blocking time for slow CPU devices.

For reference, most significant improvements derive from:

- **removing or splitting vendor libraries, and lazyloading these using `requestIdleCallback`**. On the contrary, removing home background image didn’t help much so I kept the image and just reduced its size and made it responsive. Main performance issues were related to blocking JS execution time, and largest contentful paint happening too late (it’s largest content block if there is no background image). So the priority was to optimize the sequence of requests. Here comes the next item. 
- **embedded config JSON in the app** to spare a blocking Config API call. It’s not possible to use `rel=preload` for Config since we need Authorization header with Stelace API key, so we would need to load `stelace.js` before other dependencies and make early Config (and AssetType, CustomAttributes) API calls equivalent to using `rel=preload` for static resources, due to HTTP cache header enabled for these resources on Stelace server. But it was much easier to include config JSON object in bundle.
- **responsive images** using `srcset` and new `loading="lazy"` attribute, combined with Skeleton UI, also helped to improve performance perceived by end users.

I made related UX improvements as well, such as skeleton UI, which also prevents layout shift.
Some fixes for bugs we were not aware of are also included in this PR.

![pagespeed-desktop-marketplace-demo-July2020](https://user-images.githubusercontent.com/12909094/86488243-afabbf80-bd60-11ea-82c7-27e8a0bf3ead.png)

woyuen commented on src/pages/Home.vue:
```diff
@@ -251,18 +250,53 @@ export default {
+        <picture>
+          <source
+            type="image/webp"
+            :srcset="`${
```
Is it worth to create a util to handle the responsive srcset?
```js
function getImgResponsiveSrcset (widths, getUrl) {
  return widths
    .map(width => `${getUrlFn(width) ${width}w`})
    .joins(', ')
}

// Usage:
getImgResponsiveSrcset([1024, 1366, 1600, 1920, 2560], (width) => getHomeHeroUrlTransformed({ width }))
```

abruere replied:
Yeah, DRYer indeed, but I wanted to keep breakpoints in HTML for easier authoring (closer to final HTML output).

Repetition in 2 `<source>` tags will be removed when we update our AWS image handler to new version supporting `AUTO_WEBP`.

abruere commented on src/pages/Home.vue:
```diff
@@ -164,6 +183,14 @@ export default {
       this.selectedCategory = category
     },
+    setDates ({ startDate, endDate }) {
+      if (startDate || startDate === null) {
```
Indeed, thanks.

abruere commented on scripts/include-config.js:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  const commonPath = '../src/store/common/'
+  const configPath = path.join(__dirname, `${commonPath}config-copy.json`)
```
Thanks.

abruere commented on scripts/include-config.js:
```diff
@@ -0,0 +1,89 @@
+stelace.setTimeout(5000)
+
+run()
+  .then(() => success('API Config, Asset Types and Custom Attributes copied for build'))
```
Exactly. It’s not required to load the app, but just some performance improvement.

Though a catch block should actually ensure we have some empty files at least, to avoid errors there https://github.com/stelace/marketplace-demo/blob/1c6741bffaa927695e6629ad49f3b3ec98db0582/src/store/common/state.js#L4-L6 (and `*LastFetchedDate` set to null in this store file then).

Good catch.

abruere commented on src/pages/Search.vue:
```diff
@@ -186,6 +184,9 @@ export default {
     await this.searchAssets()
     this.$store.dispatch('getHighestPrice')
+
+    if ('requestIdleCallback' in window) requestIdleCallback(this.showMapOnLargeScreen)
```
Indeed, but I was leaving this to later usage of a nice library: https://github.com/GoogleChromeLabs/idlize based on this article, which also enables more complex use cases such as queues.

abruere commented on src/pages/Home.vue:
```diff
@@ -251,18 +250,53 @@ export default {
-        @load="isBackgroundLoaded = true"
-      />
+      <div class="hero__background absolute-full">
+        <picture>
```
Exactly (transparent GIF here).

abruere commented:
I’ve just pushed temporary commits addressing your comments @woyuen.

I’ll rebase the branch later if you want to have a look at these changes first.

woyuen commented:
Thanks for the changes!
That looks good to me.
